File: unreal-plugin/Content/Python/mcp_bridge/handlers/gameplay.py
"""Gameplay handler: PIE control, telemetry, acceptance test runner.

Commands registered in router.py:
    gameplay_pie_start            -- launch PIE and wait until ready
    gameplay_pie_stop             -- end PIE session
    gameplay_telemetry_snapshot   -- capture current runtime state
    gameplay_run_acceptance_tests -- run predicates against live PIE
"""
from __future__ import annotations
import traceback
from typing import Any, Dict, List

from mcp_bridge.generation import pie_harness
from mcp_bridge.generation import telemetry_capture as tc


def handle_pie_start(params: Dict[str, Any]) -> Dict[str, Any]:
    """Launch PIE and wait until the session is ready.

    Params:
        level_path (str, optional): Content path to load before PIE.
    """
    try:
        level_path = params.get("level_path")
        ok = pie_harness.launch_pie(level_path=level_path)
        if not ok:
            return {"success": False, "data": {}, "error": "Failed to call play_in_editor()"}
        ready = pie_harness.wait_for_pie_ready()
        if not ready:
            return {
                "success": False,
                "data": {},
                "error": f"PIE did not become ready within {pie_harness._PIE_READY_TIMEOUT_S}s",
            }
        return {"success": True, "data": {"status": "pie_ready"}}
    except Exception as e:
        return {"success": False, "data": {}, "error": f"{e}\n{traceback.format_exc()}"}


def handle_pie_stop(params: Dict[str, Any]) -> Dict[str, Any]:
    """End the current PIE session."""
    try:
        ok = pie_harness.stop_pie()
        return {"success": ok, "data": {"status": "pie_stopped" if ok else "stop_failed"}}
    except Exception as e:
        return {"success": False, "data": {}, "error": str(e)}


def handle_telemetry_snapshot(params: Dict[str, Any]) -> Dict[str, Any]:
    """Capture current PIE runtime state."""
    try:
        pie_world = tc.get_pie_world()
        if pie_world is None:
            return {"success": False, "data": {}, "error": "PIE is not running"}
        frame = tc.snapshot(pie_world)
        return {
            "success": True,
            "data": {
                "log_lines": frame.log_lines_since_last,
                "possessed_pawn_class": frame.possessed_pawn_class,
                "ai_controller_states": frame.ai_controller_states,
                "visible_widgets": frame.visible_widgets,
                "fps": frame.fps,
                "pie_world_name": frame.pie_world_name,
            },
        }
    except Exception as e:
        return {"success": False, "data": {}, "error": f"{e}\n{traceback.format_exc()}"}


def handle_run_acceptance_tests(params: Dict[str, Any]) -> Dict[str, Any]:
    """Run acceptance test predicates against the live PIE session.

    Params:
        tests (List[str]): Predicate strings.
        timeout_seconds (float, optional): Per-predicate timeout override.
    """
    try:
        raw_tests: List[str] = params.get("tests", [])
        if not raw_tests:
            return {"success": False, "data": {}, "error": "No tests provided"}
        timeout_override = params.get("timeout_seconds")
        specs = pie_harness.parse_test_specs(raw_tests)
        if timeout_override is not None:
            for spec in specs:
                spec.timeout_seconds = float(timeout_override)
        results = pie_harness.run_assertions(specs)
        total = len(results)
        passed = sum(1 for r in results if r.passed)
        return {
            "success": True,
            "data": {
                "total": total,
                "passed": passed,
                "failed": total - passed,
                "results": [
                    {"predicate": r.predicate, "target": r.target, "passed": r.passed, "observed": r.observed}
                    for r in results
                ],
            },
        }
    except Exception as e:
        return {"success": False, "data": {}, "error": f"{e}\n{traceback.format_exc()}"}

File: mcp-server/src/tools/gameplay.ts
/**
 * Gameplay tools: PIE control, telemetry, acceptance tests.
 *
 * None of these commands are modifying in the undo-stack sense (PIE
 * sessions are transient). gameplay_pie_start is NOT added to modifyingCommands.
 */

import { z } from "zod";
import { UnrealClient } from "../unreal-client.js";
import type { ToolDefinition } from "../types.js";

export function createGameplayTools(client: UnrealClient): ToolDefinition[] {
  return [
    {
      name: "gameplay_pie_start",
      description:
        "Launch a Play In Editor (PIE) session and wait until the game world is ready " +
        "(up to 30s). Returns success once the PIE-ready log marker is found. " +
        "Call this before gameplay_run_acceptance_tests or gameplay_telemetry_snapshot.",
      inputSchema: z.object({
        level_path: z
          .string()
          .optional()
          .describe(
            "Content path of the level to load before PIE (e.g. /Game/Maps/Gameplay). " +
            "If omitted, uses the current level."
          ),
      }),
      handler: async (params) => {
        const result = await client.sendCommand("gameplay_pie_start", params);
        return {
          content: [{ type: "text" as const, text: JSON.stringify(result, null, 2) }],
        };
      },
    },
    {
      name: "gameplay_pie_stop",
      description: "End the current PIE session.",
      inputSchema: z.object({}),
      handler: async (params) => {
        const result = await client.sendCommand("gameplay_pie_stop", params);
        return {
          content: [{ type: "text" as const, text: JSON.stringify(result, null, 2) }],
        };
      },
    },
    {
      name: "gameplay_telemetry_snapshot",
      description:
        "Capture the current PIE runtime state: new log lines since last snapshot, " +
        "possessed pawn class, visible widget class names, AI controller states, and PIE world name. " +
        "PIE must be running.",
      inputSchema: z.object({}),
      handler: async (params) => {
        const result = await client.sendCommand("gameplay_telemetry_snapshot", params);
        return {
          content: [{ type: "text" as const, text: JSON.stringify(result, null, 2) }],
        };
      },
    },
    {
      name: "gameplay_run_acceptance_tests",
      description:
        "Run structured acceptance test predicates against a live PIE session. " +
        "PIE must already be running (call gameplay_pie_start first). " +
        "Returns pass/fail per predicate with observed value. " +
        "Predicates: 'pawn_possessed:ClassName', 'widget_visible:WidgetName', " +
        "'log_contains:String', 'ai_state:ActorName:StateName', 'survive:N' (N seconds without crash).",
      inputSchema: z.object({
        tests: z
          .array(z.string())
          .min(1)
          .describe(
            "List of predicate strings. Examples: " +
            "'pawn_possessed:BP_Character', 'widget_visible:WBP_HUD', " +
            "'log_contains:GameStarted', 'ai_state:BP_Enemy:Patrol', 'survive:5'"
          ),
        timeout_seconds: z
          .number()
          .positive()
          .optional()
          .describe("Override timeout_seconds for all predicates (default: 5.0 per predicate)"),
      }),
      handler: async (params) => {
        const result = await client.sendCommand("gameplay_run_acceptance_tests", params);
        return {
          content: [{ type: "text" as const, text: JSON.stringify(result, null, 2) }],
        };
      },
    },
  ];
}
